### PR TITLE
Fix #1279: Enable reflective instantiation via static initializers

### LINKED ISSUE #1279: Add support for portable-scala/reflect
portable-scala/reflect offers a minimalistic set of reflection primitives:

- reflectively instantiate classes that were explicitly annotated to be reachable through reflection
- reflectively instantiate modules that were explicitly annotated to be reachable through reflection

The API is meant to have the same semantics portable across all three platforms (JVM/JS/Native).

### PR DESCRIPTION
**Pull request by errikos**
This PR enables users to reflectively instantiate objects, by making use of static class initialisers (not to be confused with JVM class initialisers). Related issue is https://github.com/scala-native/scala-native/issues/1279.

To achieve the above:
- The NIR member signatures are enhanced with an extra attribute, called `clinit`, which is mangled with `I`. Static initialiser functions take no arguments and return `Unit`.
- For the static initialiser functions to be called, they must be made reachable. For this to work, we need to iterate over all program classes (each of them in a separate `.nir` file) and check whether they contain a `clinit` function. To avoid deserialising the whole file for each class, I added a single byte at the beginning of each file, to indicate whether the type it contains has a `clinit` function. If that byte is set, then the `clinit` function of the type in that file is added to the reachable set.
- When generating the code for the `main` method (native), we have to call the discovered `clinit` fuctions. We do this just before calling the main Scala method.

On top of that, we implement the same `Reflect` API and `@EnableReflectiveInstantiation`-based mechanism that Scala.js has.

Objectives list:
- [x] Add support for static initialisers to the NIR
- [x] Add the `@EnableReflectiveInstantiation` annotation
- [x] Implement the desired annotation behaviour for modules
- [x] Implement the desired annotation behaviour for normal classes

### CODE REVIEW

**Comment by ekrich** [comment]
Is this the issue in the tracker? https://github.com/scala-native/scala-native/issues/1279

**Comment by sjrd** [comment]
Yes indeed, this is for #1279.
I'll assess #1727 tomorrow.

File: nir/src/main/scala/scala/scalanative/nir/serialization/BinaryDeserializer.scala [hunk 1]
```diff
@@ -12,6 +12,8 @@ final class BinaryDeserializer(buffer: ByteBuffer) {
   private val header: Map[Global, Int] = {
     buffer.position(0)
 
+    val hasClInit = getBool
```
Comment (sjrd) [inline nir/src/main/scala/scala/scalanative/nir/serialization/BinaryDeserializer.scala]: I believe the meaning of this byte should be generalized from

> there is a static initializer in this file

to

> there is at least one reason to fully load this file even if it's not referenced elsewhere

*At the moment*, this would be equivalent to "there is a clinit", because nothing else could justify having to load the file. But if we add another feature in the IR in the future that would also require the file to be loaded, we would be able to do so without breaking backward binary compatibility. If we stick to "this file has a clinit", then we would have to add *another* byte for the new feature, breaking everything.

File: tools/src/main/scala/scala/scalanative/linker/ClassPath.scala [hunk 2]
```diff
@@ -7,6 +7,8 @@ import nir.serialization.deserializeBinary
 import java.nio.file.{FileSystems, Path}
 import scalanative.io.VirtualDirectory
 import scalanative.util.Scope
+import java.io.FileInputStream
+import java.nio.ByteBuffer
```
Comment (sjrd) [inline tools/src/main/scala/scala/scalanative/linker/ClassPath.scala]: These imports do not seem to be used (at least in this commit).

File: nir/src/main/scala/scala/scalanative/nir/serialization/BinaryDeserializer.scala (hunk 1 above)
Comment (sjrd) [reply nir/src/main/scala/scala/scalanative/nir/serialization/BinaryDeserializer.scala]: From the following commit, I believe you already have ta name for this concept: `classesWithEntryPoint`. So this could simply be renamed to `hasEntryPoint`.

File: tools/src/main/scala/scala/scalanative/linker/ClassPath.scala [hunk 3]
```diff
@@ -50,5 +54,13 @@ object ClassPath {
           deserializeBinary(directory.read(file))
         }
       })
+
+    def classesWithEntryPoints(): Iterable[Global.Top] = {
+      files.filter {
+        case (top, file) =>
+          val buffer = directory.read(file, len = 1)
+          buffer.get != 0
```
Comment (sjrd) [inline tools/src/main/scala/scala/scalanative/linker/ClassPath.scala]:
```suggestion
          buffer.get() != 0
```
`get()` is side-effecting (calling it twice will not return the same thing), so it should always be called with explicit `()`. This is a commonly used convention in Scala.

Comment (sjrd) [reply tools/src/main/scala/scala/scalanative/linker/ClassPath.scala]: This information should really be cached. Otherwise all files will have to reread on every linking request.

File: tools/src/main/scala/scala/scalanative/linker/ClassLoader.scala [hunk 4]
```diff
@@ -41,6 +47,18 @@ object ClassLoader {
...
+    def classesWithEntryPoints(): Iterable[Global] = {
+      scopes.filter {
+        case (top, defns) =>
+          defns.exists {
+            case defn: Defn.Define =>
+              val Global.Member(_, sig) = defn.name
+              sig.isClinit
+            case _ => false
+          }
+      }.keySet
```
Comment (sjrd) [inline tools/src/main/scala/scala/scalanative/linker/ClassLoader.scala]: Even in memory, I believe we should cache this result, because it is expensive to compute, and used on every linking request.

File: tools/src/main/scala/scala/scalanative/linker/Reach.scala [hunk 5]
```diff
@@ -138,6 +139,10 @@ class Reach(config: build.Config, entries: Seq[Global], loader: ClassLoader) {
             }
           }
         }
+        val clinit = cls.name.member(Sig.Clinit())
+        if (loaded(cls.name).contains(clinit)) {
+          reachGlobal(clinit)
+        }
```
Comment (sjrd) [inline tools/src/main/scala/scala/scalanative/linker/Reach.scala]: There's something fishy here. If you unroll the code above this snippet to see all of `def reachEntry`, you can see that when `reachEntry` is called, as soon as the class is not abstract, `reachAllocation(cls)` will be called, which will make instances of the class reachable. Calling `reachEntry` for classes with a static initializer will therefore cause too much methods to be reachable, because a non-abstract class with a static initializer will have its instances reachable, although the presence of a static initializer would not necessarily cause the class to be instantiated.

I believe you cannot reuse `reachEntry`. You need a different `def reachClinit` that focuses on reaching the static initializers, and not reach the allocation.

File: tools/src/main/scala/scala/scalanative/linker/Reach.scala [hunk 6]
```diff
@@ -219,6 +224,8 @@ class Reach(config: build.Config, entries: Seq[Global], loader: ClassLoader) {
                 update(Rt.JavaHashCodeSig)
               case sig if sig.isMethod || sig.isCtor || sig.isGenerated =>
                 update(sig)
+              case sig if sig.isClinit =>
+                update(sig)
```
Comment (sjrd) [inline tools/src/main/scala/scala/scalanative/linker/Reach.scala]: It looks like this could be merged with the previous `case`, just adding `|| sig.isClinit`.

File: nscplugin/src/main/scala/scala/scalanative/nscplugin/NirGenStat.scala [hunk 7]
```diff
@@ -82,6 +82,7 @@ trait NirGenStat { self: NirGenPhase =>
       val body   = cd.impl.body
 
       buf += Defn.Class(attrs, name, None, Seq.empty)
+      genReflectiveInstantiation(cd)
```
Comment (sjrd) [inline nscplugin/src/main/scala/scala/scalanative/nscplugin/NirGenStat.scala]: I don't think we should generate reflective instantiation info for structs. Or at least we should very carefully consider the consequences of doing so.

FTR, no reflective instantiation info are ever generated for JS classes in Scala.js, so there is precedent for ignoring that for interop-specific "stuff".

File: .gitignore [hunk 8]
```diff
@@ -19,3 +19,8 @@ bin/
 .settings/
 .externalToolBuilders/
 .cache*
+
+# metals
+/.bloop/
+/.metals/
+/project/metals.sbt
```
Comment (sjrd) [inline .gitignore]: This should be in a separate PR ;)

File: tools/src/main/scala/scala/scalanative/codegen/Generate.scala [hunk 9]
```diff
@@ -138,6 +138,8 @@ object Generate {
         Next.Unwind(exc, Next.Label(handler, Seq(exc)))
       }
 
+      lazy val clinitCalls = genClinitCalls(unwind)
```
Comment (sjrd) [inline tools/src/main/scala/scala/scalanative/codegen/Generate.scala]: A `lazy val` that calls a thing with by-name argument that has side effects is dangerous. Since `clinitCalls` is always called exactly once anyway, it would be easier to reason about by simply inlining `genClinitCalls(unwin)` at the only place where `clinitCalls` is used.

File: tools/src/main/scala/scala/scalanative/codegen/Generate.scala [hunk 10]
```diff
@@ -155,19 +157,33 @@ object Generate {
...
-          Inst.Let(Op.Call(entryMainTy, entryMain, Seq(module, arr)), unwind),
-          Inst.Let(Op.Call(RuntimeLoopSig, RuntimeLoop, Seq(module)), unwind),
-          Inst.Ret(Val.Int(0)),
-          Inst.Label(handler, Seq(exc)),
-          Inst.Let(Op.Call(PrintStackTraceSig, PrintStackTrace, Seq(exc)),
-                   Next.None),
-          Inst.Ret(Val.Int(1))
+                   unwind)
         )
+          ++ clinitCalls // call the class initializers before calling main
```
Comment (sjrd) [inline tools/src/main/scala/scala/scalanative/codegen/Generate.scala]: Wouldn't it be better to insert this between the `Let`s for `rt` and `arr`? That would avoid the latter two's lifetimes to extend over the calls to the static initializers.

**Comment by sjrd** [comment]
For tests, you should be able to just copy https://github.com/scala-js/scala-js/blob/3534fda873127e5da1badbefa376ce0c7d951d03/test-suite/js/src/test/scala/org/scalajs/testsuite/library/ReflectTest.scala and adapt the package names and imports. Everything should work the same as in Scala.js.

File: nscplugin/src/main/scala/scala/scalanative/nscplugin/NirGenStat.scala [hunk 11]
```diff
@@ -218,6 +229,354 @@ trait NirGenStat { self: NirGenPhase =>
...
+                  // If the expected argument type can be boxed (i.e. is a primitive
+                  // type), then we need to unbox it before passing it to C.
+                  Type.box.get(arg) match {
+                    case Some(_) =>
+                      // first, cast the primitive type to java.lang.Number
+                      val num = exprBuf.as(jlNumberRef, elem, unwind(curFresh))
+                      // then, convert to the desired type
+                      val conv = exprBuf.method(num,
+                                                Type.primConvSig(arg),
+                                                unwind(curFresh))
```
[the remark this answers is not in the document]
Comment (errikos) [reply nscplugin/src/main/scala/scala/scalanative/nscplugin/NirGenStat.scala:467]: I think that we also need to take care of shrinking conversions. If a constructor expects a `Short` and someone provides e.g. `42`, which is an `Int`, we still have a mismatch.

I think that we need something more generic, like:

```scala
  private def primConversionIfRequired(arg: Any,
                                       paramType: Class[_]): Any = {
    paramType match {
      case java.lang.Byte.TYPE =>
        arg.asInstanceOf[Number].byteValue()
      case java.lang.Short.TYPE =>
        arg.asInstanceOf[Number].shortValue()
      case java.lang.Integer.TYPE =>
        arg.asInstanceOf[Number].intValue()
      case java.lang.Long.TYPE =>
        arg.asInstanceOf[Number].longValue()
      case java.lang.Float.TYPE =>
        arg.asInstanceOf[Number].floatValue()
      case java.lang.Double.TYPE =>
        arg.asInstanceOf[Number].doubleValue()
      case _ => arg
    }
  }
```
It is basically what I was doing in the codegen, moved to the library side.

[the remark this answers is not in the document]
Comment (errikos) [reply nscplugin/src/main/scala/scala/scalanative/nscplugin/NirGenStat.scala:467]: > This won't work for booleans and characters.

Why do we need to convert `Boolean`s and `Character`s? They are not numeric types.

[the remark this answers is not in the document]
Comment (sjrd) [reply nscplugin/src/main/scala/scala/scalanative/nscplugin/NirGenStat.scala:467]: If you follow the link to the spec I gave, you'll see that narrowing conversions shouldn't be accepted, and that characters can be converted to `int` and bigger.

As validation, here is a transcript of a REPL on the JVM to show that this is indeed what happens:
```scala
scala> case class Foo(x: Int, y: java.lang.Long)
defined class Foo

scala> val c = classOf[Foo].getDeclaredConstructors()(0)
c: java.lang.reflect.Constructor[_] = public Foo(int,java.lang.Long)

scala> c.newInstance(1, 2L)
res4: Object = Foo(1,2)

scala> c.newInstance(1, 2)
java.lang.IllegalArgumentException: argument type mismatch
  at sun.reflect.NativeConstructorAccessorImpl.newInstance0(Native Method)
  at sun.reflect.NativeConstructorAccessorImpl.newInstance(NativeConstructorAccessorImpl.java:62)
  at sun.reflect.DelegatingConstructorAccessorImpl.newInstance(DelegatingConstructorAccessorImpl.java:45)
  at java.lang.reflect.Constructor.newInstance(Constructor.java:423)
  ... 34 elided

scala> c.newInstance(1, 2.toShort)
java.lang.IllegalArgumentException: argument type mismatch
  at sun.reflect.NativeConstructorAccessorImpl.newInstance0(Native Method)
  at sun.reflect.NativeConstructorAccessorImpl.newInstance(NativeConstructorAccessorImpl.java:62)
  at sun.reflect.DelegatingConstructorAccessorImpl.newInstance(DelegatingConstructorAccessorImpl.java:45)
  at java.lang.reflect.Constructor.newInstance(Constructor.java:423)
  ... 34 elided

scala> c.newInstance(1L, 2L)
java.lang.IllegalArgumentException: argument type mismatch
  at sun.reflect.NativeConstructorAccessorImpl.newInstance0(Native Method)
  at sun.reflect.NativeConstructorAccessorImpl.newInstance(NativeConstructorAccessorImpl.java:62)
  at sun.reflect.DelegatingConstructorAccessorImpl.newInstance(DelegatingConstructorAccessorImpl.java:45)
  at java.lang.reflect.Constructor.newInstance(Constructor.java:423)
  ... 34 elided

scala> c.newInstance(1.toShort, 2L)
res8: Object = Foo(1,2)

scala> c.newInstance(1.4, 2L)
java.lang.IllegalArgumentException: argument type mismatch
  at sun.reflect.NativeConstructorAccessorImpl.newInstance0(Native Method)
  at sun.reflect.NativeConstructorAccessorImpl.newInstance(NativeConstructorAccessorImpl.java:62)
  at sun.reflect.DelegatingConstructorAccessorImpl.newInstance(DelegatingConstructorAccessorImpl.java:45)
  at java.lang.reflect.Constructor.newInstance(Constructor.java:423)
  ... 34 elided

scala> c.newInstance('A', 2L)
res10: Object = Foo(65,2)

scala> case class Foo(x: Int, y: java.lang.Long, z: Short, c: Char)
defined class Foo

scala> val c = classOf[Foo].getDeclaredConstructors()(0)
c: java.lang.reflect.Constructor[_] = public Foo(int,java.lang.Long,short,char)

scala> c.newInstance('A', 2L, 'A', 'B')
java.lang.IllegalArgumentException: argument type mismatch
  at sun.reflect.NativeConstructorAccessorImpl.newInstance0(Native Method)
  at sun.reflect.NativeConstructorAccessorImpl.newInstance(NativeConstructorAccessorImpl.java:62)
  at sun.reflect.DelegatingConstructorAccessorImpl.newInstance(DelegatingConstructorAccessorImpl.java:45)
  at java.lang.reflect.Constructor.newInstance(Constructor.java:423)
  ... 34 elided

scala> c.newInstance('A', 2L, 5, 'B')
java.lang.IllegalArgumentException: argument type mismatch
  at sun.reflect.NativeConstructorAccessorImpl.newInstance0(Native Method)
  at sun.reflect.NativeConstructorAccessorImpl.newInstance(NativeConstructorAccessorImpl.java:62)
  at sun.reflect.DelegatingConstructorAccessorImpl.newInstance(DelegatingConstructorAccessorImpl.java:45)
  at java.lang.reflect.Constructor.newInstance(Constructor.java:423)
  ... 34 elided

scala> c.newInstance('A', 2L, 5.toByte, 'B')
res13: Object = Foo(65,2,5,B)

scala> c.newInstance('A', 2L, 5.toShort, 'B')
res14: Object = Foo(65,2,5,B)

scala> c.newInstance('A', 2L, 5.toShort, 5.toShort)
java.lang.IllegalArgumentException: argument type mismatch
  at sun.reflect.NativeConstructorAccessorImpl.newInstance0(Native Method)
  at sun.reflect.NativeConstructorAccessorImpl.newInstance(NativeConstructorAccessorImpl.java:62)
  at sun.reflect.DelegatingConstructorAccessorImpl.newInstance(DelegatingConstructorAccessorImpl.java:45)
  at java.lang.reflect.Constructor.newInstance(Constructor.java:423)
  ... 34 elided

scala> c.newInstance('A', 2L, 5.toShort, 5.toByte)
java.lang.IllegalArgumentException: argument type mismatch
  at sun.reflect.NativeConstructorAccessorImpl.newInstance0(Native Method)
  at sun.reflect.NativeConstructorAccessorImpl.newInstance(NativeConstructorAccessorImpl.java:62)
  at sun.reflect.DelegatingConstructorAccessorImpl.newInstance(DelegatingConstructorAccessorImpl.java:45)
  at java.lang.reflect.Constructor.newInstance(Constructor.java:423)
  ... 34 elided
```

[the remark this answers is not in the document]
Comment (sjrd) [reply nscplugin/src/main/scala/scala/scalanative/nscplugin/NirGenStat.scala:467]: > 
> 
> > This won't work for booleans and characters.
> 
> Why do we need to convert `Boolean`s and `Character`s? They are not numeric types.

`Boolean`s need not be converted, but the code that is currently in the PR would try to cast `Boolean`s to `Number`s anyway, even if the expected target type was `Boolean`. Or maybe I misunderstood the code. For `Char` we in fact need to convert them to numeric types, as I showed above.

[the remark this answers is not in the document]
Comment (errikos) [reply nscplugin/src/main/scala/scala/scalanative/nscplugin/NirGenStat.scala:467]: Okay, thanks for the clarification.

I got confused from a test case from Scala.js. In the `testClassCtorWithArgs` test, the `Short` constructor is requested and I though it was "converted" from the `Int` one. However I now see the presence of the `VC` class. The current implementation does not support such a scenario.

[the remark this answers is not in the document]
Comment (sjrd) [reply nscplugin/src/main/scala/scala/scalanative/nscplugin/NirGenStat.scala:467]: The current implementation in Scala.js seems incomplete about the conversions. We'll need to fix that when I get the chance.

[the remark this answers is not in the document]
Comment (errikos) [reply nscplugin/src/main/scala/scala/scalanative/nscplugin/NirGenStat.scala:467]: So, if I understand this correctly, instantiating a `Short` constructor with `42`, which is [supported](https://github.com/scala-js/scala-js/blob/3534fda873127e5da1badbefa376ce0c7d951d03/test-suite/js/src/test/scala/org/scalajs/testsuite/library/ReflectTest.scala#L179) in Scala.js, should not be supported in Native.

[the remark this answers is not in the document]
Comment (errikos) [reply nscplugin/src/main/scala/scala/scalanative/nscplugin/NirGenStat.scala:467]: Therefore, if I understand this correctly, calling a `Short` constructor with a value of `42` (which is an `Int`), which is [supported by Scala.js](https://github.com/scala-js/scala-js/blob/3534fda873127e5da1badbefa376ce0c7d951d03/test-suite/js/src/test/scala/org/scalajs/testsuite/library/ReflectTest.scala#L179), should not be supported by Native?

I suppose that Scala.js treats numeric values differently, in a JS-specific way. I am a little confused as to which semantics we would like to "follow". In C for example, it is OK (to some extent) to do all kinds of conversions. Why not do the same in Native and why is the behaviour of Scala.js different in that aspect?

[the remark this answers is not in the document]
Comment (sjrd) [reply nscplugin/src/main/scala/scala/scalanative/nscplugin/NirGenStat.scala]: > Therefore, if I understand this correctly, calling a Short constructor with a value of 42 (which is an Int), which is supported by Scala.js, should not be supported by Native?

That's right.

> I suppose that Scala.js treats numeric values differently, in a JS-specific way. I am a little confused as to which semantics we would like to "follow". In C for example, it is OK (to some extent) to do all kinds of conversions. Why not do the same in Native and why is the behaviour of Scala.js different in that aspect?

We have to follow the JVM semantics first. Copying C or Scala.js is secondary.

Scala.js treats boxed primitive values differently than the JVM across the board (see https://www.scala-js.org/doc/semantics.html) for JS-specific reasons, and is therefore not a gold standard to follow if the JVM does not agree. Basically in Scala.js we *cannot* tell apart a `j.l.Integer` with value 42 from a `j.l.Integer` with value 42, so we're forced to accept it.

> On the other hand, Scala seems to accept the above scenario:

That's with *compile-time* adaptation. Java does the same in that situation. But here we are concerned with *run-time* adaptation as performed by Java reflection, since that is the capability we are trying to emulate.

File: nscplugin/src/main/scala/scala/scalanative/nscplugin/NirGenSymbols.scala [hunk 12]
```diff
@@ -0,0 +1,25 @@
...
+  val serializable = Global.Top("scala.Serializable")
+
+  val jlClass    = Global.Top("java.lang.Class")
+  val jlClassRef = Type.Ref(jlClass)
+
+  val jlObject    = Global.Top("java.lang.Object")
+  val jlObjectRef = Type.Ref(jlObject)
+
+  val jlNumber    = Global.Top("java.lang.Number")
+  val jlNumberRef = Type.Ref(jlNumber)
```
Comment (sjrd) [inline nscplugin/src/main/scala/scala/scalanative/nscplugin/NirGenSymbols.scala:16]: This is now unused, I think.

**Comment by errikos** [comment]
From what I understand, this is going to be squashed and merged with the first comment being the commit message. Will edit today with a summary of the changes.

**Comment by sjrd** [comment]
Oops, cross-action. I took the PR description to make the squashed commit message. It contained the most important info.

**Comment by errikos** [comment]
I would add a few more details, but it was good enough. Thanks!
